core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1405,11 +1393,28 @@ class ReplicaManager(val config: KafkaConfig,
                   s"leader epoch $currentLeaderEpoch")
                 responseMap.put(topicPartition, Errors.STALE_CONTROLLER_EPOCH)
               } else {
-                stateChangeLogger.info(s"Ignoring LeaderAndIsr request from " +
-                  s"controller $controllerId with correlation id $correlationId " +
-                  s"epoch $controllerEpoch for partition $topicPartition since its associated " +
-                  s"leader epoch $requestLeaderEpoch matches the current leader epoch")
-                responseMap.put(topicPartition, Errors.STALE_CONTROLLER_EPOCH)
+                // The controller may send LeaderAndIsr to upgrade to using topic IDs without bumping the epoch.
+                val error = requestTopicId match {
+                  case Some(topicId) if partition.log.isEmpty =>
+                    // We wanted to assign the topic ID to the log, but the log was unavailable.
+                    stateChangeLogger.info(s"Tried to update log for $topicPartition to assign topic ID " +
+                      s"$topicId from LeaderAndIsr request from controller $controllerId with correlation" +
+                      s" id $correlationId epoch $controllerEpoch")
+                    Errors.KAFKA_STORAGE_ERROR
```

Comment:
Hmm, I am not sure about this. If you look at `localLogOrException`, for example, it would return `NotLeader` if the log is not defined. In this method, we already have the `replicaStateChangeLock`, so that protects us from a concurrent call to `markPartitionOffline`. As far as I can tell, the only time we might not have a `Log` defined for the partition is if it has not been created yet or if it has been deleted. But given that we have already verified that the epoch of the request matches the local epoch, neither case seems possible. My suggestion would be to remove this check and use `localLogOrException` to make it clear that the log is expected to exist at this point.

Reply:
I'm not sure I follow here. The result of localLogOrException in makeLeader/makeFollower is catching the NotLeader to return KAFKA_STORAGE_ERROR. The only difference is the throwing of the error. 

If we were to use localLogOrException, we would still need to catch it and return KAFKA_STORAGE_ERROR. I can do that if it makes more sense. 

I believe the same about the epoch matching, and think the only case where we don't have log is when we have a storage error.

Reply:
Ah hmmm maybe I actually misread this code previously. I'll take a look again.